src/main/java/org/springframework/integration/aws/outbound/KinesisMessageHandler.java

```diff
@@ -163,19 +211,21 @@ protected void onInit() throws Exception {
 
 	@Override
 	@SuppressWarnings("unchecked")
-	protected void handleMessageInternal(Message<?> message) throws Exception {
+	protected void handleMessageInternal(final Message<?> message) throws Exception {
 		Future<?> resultFuture = null;
+
 		if (message.getPayload() instanceof PutRecordsRequest) {
+
 			resultFuture = this.amazonKinesis.putRecordsAsync((PutRecordsRequest) message.getPayload(),
-					(AsyncHandler<PutRecordsRequest, PutRecordsResult>) this.asyncHandler);
+					getPutRecordsAsyncHandler(message));
 		}
 		else {
-			PutRecordRequest putRecordRequest = (message.getPayload() instanceof PutRecordRequest)
+			final PutRecordRequest putRecordRequest = (message.getPayload() instanceof PutRecordRequest)
```

Comment:
I know this is existing code, but looks like it is wrong a bit and can be combined with the logic in the `if` before.
I mean we don't need to have two `this.amazonKinesis.putRecordAsync()` call and don't need to check for the `PutRecordRequest` type again.

Reply:
I'm not sure I understand what you mean. One call is `#putRecordsAsync` while the other is `#putRecordAsync` and they require different parameter types depending on the payload. The payload of the `Message` can be both `PutRecordsRequest` and `PutRecordRequest` (among others) so we need those checks?

Reply:
Ah! I see your point.
Yeah.. bad vision 😄 

So, yes, you are right they are both about different matters.

However I think we still can have a generic `getAsyncHandler(final Message<?> message,
			final AmazonWebServiceRequest request)`

And only check a difference when we build a `Message` in the `onSuccess()` since `PutRecordsResult` is really different from the `PutRecordResult`.

Thanks

WDYT?

Reply:
Hah same thing with me right away  😄 

I've worked out a fairly generic `#getAsyncHandler` that I’ve just pushed. I'm not sure how happy I am with it so I'll be looking forward to your input. Thanks for your time reviewing!

Reply:
Well, good attempt, indeed. Let’s consider if we can make this new method really generic - with the `<T>` type.